pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+		},
+
+		/// Emitted when a new bid is placed on an auction.
+		Bid { who: T::AccountId, auction_id: AuctionId, price: T::Balance },
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_auction)]
+	pub type NextAuctionId<T: Config> = StorageValue<_, AuctionId, ValueQuery>;
+
+	/// Stores the data associated with each auction.
+	#[pallet::storage]
+	#[pallet::getter(fn auctions)]
+	pub type Auctions<T: Config> = StorageMap<_, Blake2_128Concat, AuctionId, AuctionData<T>>;
+
+	/// Stores the auction IDs that are set to expire at a specific block.
+	#[pallet::storage]
+	pub type AuctionsExpiryBlock<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, Vec<AuctionId>, ValueQuery>;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(block_number: BlockNumberFor<T>) {
+			// Expire auctions that are due.
+			AuctionsExpiryBlock::<T>::take(block_number).into_iter().for_each(|auction_id| {
+				if let Some(auction_data) = Auctions::<T>::take(auction_id) {
+					Self::resolve_auction(auction_id, auction_data);
+				}
+			});
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates a new auction for a specified NFT.
+		/// This function can only be called by an account with the Customer role.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::create_auction())]
+		pub fn create_auction(
+			origin: OriginFor<T>,
+			nft_id: NftId,
+			start: Option<T::Balance>,
+			reserve: Option<T::Balance>,
+			buy_now: Option<T::Balance>,
+		) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// check the NftIdState is Free
+			T::NftManager::ensure_nft_state(nft_id, NftState::Free)?;
+
+			// Check the Nft is belong to the origin
+			T::NftManager::ensure_nft_is_valid(&id, nft_id)?;
+
+			// Set the Auctions storage
+			let auction_id = Self::next_auction_id();
+			let expiry_block =
+				T::AuctionLength::get() + frame_system::Pallet::<T>::current_block_number();
+
+			Auctions::<T>::insert(
+				auction_id,
+				AuctionData {
+					nft_id,
+					owner: id.clone(),
+					start,
+					reserve,
+					buy_now,
+					expiry_block,
+					current_bid: None,
+				},
+			);
+
+			// Append the auction id to the AuctionsExpiryBlock storage.
+			AuctionsExpiryBlock::<T>::append(expiry_block, auction_id);
+
+			// Change Nft state to auction.
+			T::NftManager::change_nft_state(nft_id, NftState::Auction(auction_id))?;
+
+			// Pay fee.
+			T::Bank::transfer(&id, &T::Bank::treasury()?, T::AuctionStartFee::get())?;
+
+			Self::deposit_event(Event::<T>::AuctionCreated { who: id, auction_id });
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::bid())]
+		pub fn bid(
+			origin: OriginFor<T>,
+			auction_id: AuctionId,
+			price: T::Balance,
```

Comment:
the variable names need to be improved. there are 2 sets of bids: 
(current_bidder, current_bid)
(new_bidder, new_bid)
the names (whatever you choose to use for the 2) must be clear and distinguishable.